### ISSUE TITLE
Cannot 'resolveSync' because the fileSystem is not sync

### ISSUE THREAD

**Issue by jantimon** [callstack/linaria#845]
## Environment

```
@linaria/webpack5-loader/node_modules/enhanced-resolve/lib/Resolver.js:242
                        throw new Error(
                        ^
Error: Cannot 'resolveSync' because the fileSystem is not sync. Use 'resolve'!
```

- Linaria version: "3.0.0-beta.13",
- Bundler (+ version): next": "11.0.1" (with webpack 5)
- Node.js version:  v14.16.1
- OS: OS-X

## Description

Can't get linaria up and running as the @linaria/webpack-loader throws an error:

```

Error: /my-project/src/foo.tsx: An unexpected runtime error occurred during dependencies evaluation: 
/my-project/node_modules/@linaria/webpack5-loader/node_modules/enhanced-resolve/lib/Resolver.js:242
                        throw new Error(
                        ^

Error: Cannot 'resolveSync' because the fileSystem is not sync. Use 'resolve'!
It may happen when your code or third party module is invalid or uses identifiers not available in Node environment, eg. window. 
Note that line numbers in above stack trace will most likely not match, because Linaria needed to transform your code a bit.

```

## Reproducible Demo

Here is a reproduction of the bug: [bug.zip](https://github.com/callstack/linaria/files/7232118/bug.zip)

## related issues:

- https://github.com/babel/babel/issues/13798
- https://github.com/vercel/next.js/issues/29410

**Comment by J-env** [comment]
I had the same problem.

next.js: 11.1.2, 11.1.0 (with webpack 5, with webpack 4)

**Comment by jantimon** [comment]
unfortunately there is no code comment explaining why the loader needs to run sync

**Comment by jantimon** [comment]
Here is a reproduction of the bug: [bug.zip](https://github.com/callstack/linaria/files/7232118/bug.zip)

It is caused by a Resolve plugin which gets injected automatically by the [jsconfig-paths-plugin](https://github.com/vercel/next.js/blob/0af3b526408bae26d6b3f8cab75c4229998bf7cb/packages/next/build/webpack/plugins/jsconfig-paths-plugin.ts) as soon as you specify paths inside `tsconfig.js` to resolve dependencies

**Comment by J-env** [comment]
@jantimon 
Manual test `/node_modules/@linaria/webpack5-loader` replace with `@linaria/webpack-loader 3.0.0-beta.7`([webpack5-loader.zip](https://github.com/callstack/linaria/files/7232929/webpack5-loader.zip)) solves this problem. 

but `npm i -D @linaria/webpack-loader@3.0.0-beta.7 ` cannot install the specified version

**Comment by jantimon** [comment]
@J-env webpack-loader is only a small helper to load either webpack4-loader or webpack5-loader

the resolve code of linaria is not the problem - linaria is using webpacks resolving logic. unfortunately the webpack resolving logic is sometimes sync and sometimes async depending on the webpack config.

next js ships with a webpack config which will be sync by default but async as soon as tsconfig includes a [paths mapping](https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping)

as soon as enhanced-require detects an async resolution hook it throws

**Comment by Anber** [comment]
Thank you @jantimon and @J-env 
This is a regression from https://github.com/callstack/linaria/pull/830. I don't see any solution for that case. I have to think.

**Comment by jantimon** [comment]
@Anber I managed to fix it on nextjs side: https://github.com/vercel/next.js/discussions/29421

I also found a solution which is probably better than the current one as it has the following advantages:

- comes with integrated caching (no need to maintain linaria-cache and cache invalidation anymore)
- works with sync & async path resolvers
- doesn't require deprecated webpack apis

however it comes also with one big disadvantage:

- it's only supported for webpack 5.32 or newer
- would take some effort to migrate the current code

the solution would be to use webpacks [this.importModule](https://webpack.js.org/api/loaders/#thisimportmodule) loader api which was developed exactly for cases like linaria

**Comment by Anber** [comment]
@jantimon that's an interesting proposal!

> it's only supported for webpack 5

It's not a problem, since we have different packages for v4 and v5.

> would take some effort to migrate the current code

It can be a problem, but maybe we can use those changes in other loaders too.

**Comment by jantimon** [comment]
Cool :) I'll try to find some time for a POC this week

**Comment by jantimon** [comment]
I have a first (very very limited) POC up and running 🎉
Right now it is blocked by a bug in css-loader: https://github.com/webpack-contrib/css-loader/issues/1381

If you like we can have a short call next week to take a look :)

**Comment by jantimon** [comment]
I found a workaround for now but I'll also file a webpack bug

**Comment by Anber** [comment]
So… is it still possible to implement a fix for webpack 5?

**Comment by Anber** [comment]
I've just added a fallback to the default configuration if an async plugin is detected. I hope it will allow releasing 3.0 without introducing significant changes suggested by @jantimon. However,  I'm still interested in a better solution. @jantimon, can you open a PR with the implementation of your approach?

**Comment by jantimon** [comment]
It's rather a POC than a working PR.. but right now I set it on hold because Vercel is moving away from Babel..  
This means using Linaria for NextJs will slow down the build by upto 10 times 😕

**Comment by morriq** [comment]
I can confirm nextjs 12 works with linaria @ beta

**Comment by Anber** [comment]
Ok, let's close this issue for now. Maybe I'll find time for supporting SWC.
